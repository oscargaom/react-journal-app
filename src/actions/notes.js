import { db } from "../firebase/firebase-config";
import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";
import { fileUpload } from "../helpers/fileUpload";

import Swal from "sweetalert2";

export const startNewNote = () => {
    /*  También podemos obtener el estado a través de una función que es enviada 
        por thunk como parámetro junto con el dispatch. Gracias a ello puedemos 
        tener acceso a todo el state y a los reducers y sería equivalente a usar
        el hook useSelector.
    */
    return async (dispatch, getState) => {
        // console.log(getState());
        const { auth: { uid } } = getState();
        // console.log(uid);

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        };

        const doc = await db.collection(`${uid}/journal/notes`).add(newNote);

        dispatch(activeNote(doc.id, newNote));
    };
};

export const activeNote = (id, note) => ({
    type: types.notesActive,
    payload: {
        id,
        ...note
    }
});


export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        // console.log(notes);
        dispatch(setNotes(notes));
    };
};


export const startSaveNote = (note) => {
    return async (dispatch, getState) => {

        const { uid } = getState().auth;

        // console.log('startSaveNote note');
        // console.log(note);
        const noteToUpdate = { ...note };
        // console.log(noteToUpdate);
        delete noteToUpdate.id;
        // console.log(noteToUpdate);

        if (!noteToUpdate.url) {
            delete noteToUpdate.url;
        }

        try {
            await db.doc(`${uid}/journal/notes/${note.id}`).update(noteToUpdate);
        } catch (error) {
            Swal.fire('Cant saved your note', error.toString(), 'error')
        }

        // console.log('startSaveNote note.id');
        // console.log(note.id);
        dispatch(refreshNote(note.id, note));

        Swal.fire('Saved note', 'Your note has been saved', 'success');
    };
};

export const setNotes = (notes) => {
    return {
        type: types.notesLoad,
        payload: notes
    };
};

export const refreshNote = (id, note) => {
    return {
        type: types.notesUpdated,
        payload: note
    };
};

export const startUploading = (file) => {
    return async (dispatch, getState) => {

        const { notes: { active } } = getState();

        // console.log(file);
        // console.log(active);

        Swal.fire({
            title: 'Uploading ...',
            text: 'Please wait ...',
            allowOutsideClick: false,
            onBeforeOpen: () => {
                Swal.showLoading();
            }
        });

        const fileUrl = await fileUpload(file);
        active.url = fileUrl;

        // console.log(fileUrl);

        dispatch(startSaveNote(active));

        Swal.close();
    }
}

export const startDeleting = (id) => {
    return async (dispatch, getState) => {
        const { auth: { uid } } = getState();

        // console.log(`startDeleting uid: ${uid}`);

        await db.doc(`${uid}/journal/notes/${id}`).delete();

        dispatch(deleteNote(id));
    }
}

export const deleteNote = (id) => ({
    type: types.notesDelete,
    payload: id
})


export const cleanNotes = () => ({
    type: types.notesLogoutCleaning,
    payload: null
})