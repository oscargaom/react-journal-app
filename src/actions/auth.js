import { types } from "../types/types";
import Swal from 'sweetalert2'
import { firebase, googleAuthProvider } from '../firebase/firebase-config'
import { startLoading, finishLoading } from "./ui";
import { cleanNotes } from "./notes";

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({ user }) => {
                const { displayName, uid } = user;
                // console.log(displayName);
                // console.log(uid);
                dispatch(login(displayName, uid));
            });
    };
};

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        // console.log(email);
        // console.log(password);
        // console.log(name);

        firebase.auth().createUserWithEmailAndPassword(email, password)
            .then(async (userCredential) => {
                // console.log(userCredential);
                const { user } = userCredential;
                await user.updateProfile({ displayName: name });
                const { displayName, uid } = user;
                // console.log(displayName);
                // console.log(uid);

                // console.log(user);
                // console.log(displayName);
                Swal.fire('Bienvenido', 'Tu cuenta ha sido creada', 'success');
                dispatch(login(uid, displayName));

            })
            .catch(error => {
                // console.log(error);
                Swal.fire('Register error', error.message, 'error')
            });
    };
}

/*  Esta es una acción asíncrona, por ello en vez de regresar un objeto regresa 
    un callback el cual el middleware thunk queda en espera de la ejecución de algún
    dispatch; en caso de que no se ejecutará algún dispatch simplmente termina. 
    El dispatch es proporcionado por el middleware thunk, así es que no debemos 
    preocuparnos por obtenerlo de algún otro lugar.
*/
export const startLoginEmailPassword = (email, password) => {
    return (dispatch) => {

        // setTimeout(() => {
        //     dispatch(login('Gpe', '456789'));
        // }, 3000);


        dispatch(startLoading());

        // setTimeout(() => {
        return firebase.auth().signInWithEmailAndPassword(email, password)
            .then((userCredential) => {
                dispatch(finishLoading());
                const { user: { displayName, uid } } = userCredential;
                Swal.fire('Bienvenido',`Hola ${displayName}`,'success');
                dispatch(login(uid, displayName));
            })
            .catch(error => {
                // const {code} = error;
                dispatch(finishLoading());
                dispatch(login('', ''));
                Swal.fire('Login Error', error.message, 'error');
            });
        // }, 9000);
    };
};

export const login = (uid, displayName) => ({
    type: types.login,
    payload: {
        uid,
        displayName
    }
});

export const logout = () => ({
    type: types.logout
});

export const startLogout = () => {
    return async (dispatch) => {
        await firebase.auth().signOut();

        dispatch(logout());

        dispatch(cleanNotes());
    };
}



