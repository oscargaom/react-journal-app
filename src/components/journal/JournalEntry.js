import React from 'react'
import moment from 'moment';
import { useDispatch } from 'react-redux';
import { activeNote } from '../../actions/notes';

export const JournalEntry = ({ note }) => {

    // console.log(note);
        
    const {id, body, title, url, date} = note;
    
    const noteDate = moment(date);

    const dispatch = useDispatch();

    const handleEntryClick = () => {
        // console.log('handleEntryClick');
        dispatch(activeNote(id, note));
    };

    return (
        <div
            className="journal__entry pointer animate__animated animate__backInLeft animate__faster"
            onClick={handleEntryClick}
        >
            <div
                className="journal__entry-picture"
                style={{
                    backgroundSize: 'cover',
                    backgroundImage: `url(${url})`
                }}
            >
            </div>

            <div className="journal__entry-body">
                <p className="journal__entry-title">
                    {title}
                </p>
                <p className="journal__entry-content">
                    {body}
                </p>
            </div>

            <div className="journal__entry-date-box">
                <span>{noteDate.format('dddd')}</span>
                <h4>{noteDate.format('Do')}</h4>
            </div>
        </div>
    )
}
