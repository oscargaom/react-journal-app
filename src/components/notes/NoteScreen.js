import React, { useEffect, useRef } from 'react'
import { NotesAppBar } from './NotesAppBar'
import { useForm } from '../../hooks/useForm'
import { useSelector, useDispatch } from 'react-redux';
import { activeNote, startDeleting } from '../../actions/notes';

export const NoteScreen = () => {

    const dispatch = useDispatch();

    const { notes: { active: note } } = useSelector(state => state);

    // console.log(note);
    const [formValues, formHandleInputChange, reset] = useForm(note);

    const { body, title, url , id} = note;
    
    /*  useRef almacena una variable mutable que en este caso es note.id la 
        cual no va a redibujar a todo el componente NoteScreen si esta cambia.
    */
    const noteId = useRef(note.id);
    // console.log(noteId);
    
    useEffect(() => {
        /*  Como useForm utiliza su propio state debemos modificar el valor de
            dicho estado para que sea redibujado el componente con los valores
            de la nota, ya que de lo contrario siempre muestra el primer valor.
        */
        if (noteId.current !== note.id) {
            reset(note);
            noteId.current = note.id;
        }
    }, [reset, note]);

    
    useEffect(() => {
        // console.log(formValues);
        dispatch(activeNote(formValues.id, {...formValues}));
    }, [formValues, dispatch])
    
    const handleDelete = () => {
        // console.log(id);
        dispatch(startDeleting(id));
    }

    return (
        <div className="notes__main-content">
            <NotesAppBar />

            <div className="notes__content">
                <input
                    type="text"
                    name="title"
                    placeholder="Some awesome title"
                    className="notes__title-input"
                    autoComplete="off"
                    value={title}
                    onChange={formHandleInputChange}
                />

                <textarea
                    name="body"
                    placeholder="What happened today"
                    className="notes__textarea"
                    value={body}
                    onChange={formHandleInputChange}
                >
                </textarea>

                {
                    (url)
                    &&
                    (<div className="notes__image">
                        <img
                            src={url}
                            alt="imagen"
                        />
                    </div>)
                }
            </div>

            <button
                className="btn btn-danger"
                onClick={handleDelete}
            >
                Delete
            </button>
        </div>
    )
}
