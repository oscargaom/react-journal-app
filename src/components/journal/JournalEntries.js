import React from 'react'
import { JournalEntry } from './JournalEntry';
import { useSelector } from 'react-redux';

export const JournalEntries = () => {

    //const entries =  [1,2,3,4,5];
    //const entries =  [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15];

    const stateSelector = useSelector(state => state);
    const { notes: entries } = stateSelector.notes;
    // console.log(entries);

    return (
        <div className="journal__entries">
            {
                entries && entries.map(note => (
                    <JournalEntry 
                        key={note.id} 
                        note={note}
                    />
                ))
            }
        </div>
    )
}
