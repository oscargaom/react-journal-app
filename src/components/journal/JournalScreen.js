import React from 'react'
import { Sidebar } from './Sidebar'
import { NoteScreen } from '../notes/NoteScreen'
import { NothingSelected } from './NothingSelected'
import { useSelector } from 'react-redux'

export const JournalScreen = () => {

    const stateSelector = useSelector(state => state);

    const { notes: { active } } = stateSelector;

    return (
        <div className="journal__main-content">

            <Sidebar />

            <main>
                {
                    (active)
                        ? <NoteScreen />
                        : <NothingSelected />
                }
            </main>
        </div>
    )
}
