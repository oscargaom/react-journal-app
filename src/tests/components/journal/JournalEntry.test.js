import React from 'react'
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import { JournalEntry } from '../../../components/journal/JournalEntry';
import { initialStateAuthNotesActiveNote } from '../../utils/initialState';
import { Provider } from 'react-redux';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { types } from '../../../types/types';
import { activeNote } from '../../../actions/notes';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

describe('Pruebas en el componente <JournalEntry />', () => {

    let store = mockStore(initialStateAuthNotesActiveNote);

    beforeEach(() => {
        store = mockStore(initialStateAuthNotesActiveNote);
        store.dispatch = jest.fn();
    });

    test('debe mostrar el snapshot correctamente ', () => {

        const { active } = initialStateAuthNotesActiveNote.notes;

        const wrapper = mount(
            <Provider store={store}>
                <JournalEntry note={active} />
            </Provider>
        );

        expect(wrapper).toMatchSnapshot();
    });

    test('debe de ejecutarse el dispatch en activeNote al ser activada la nota', () => {

        const { active } = initialStateAuthNotesActiveNote.notes;

        const wrapper = mount(
            <Provider store={store}>
                <JournalEntry note={active} />
            </Provider>
        );

        wrapper.find('.journal__entry').prop('onClick')();

        expect(store.dispatch).toHaveBeenCalledTimes(1);
        expect(store.dispatch).toHaveBeenCalledWith({
            type: types.notesActive,
            payload: active
        });
        //active.body = '4242342423';
        expect(store.dispatch).toHaveBeenLastCalledWith(
            activeNote(active.id, active)
        );
    });
})


