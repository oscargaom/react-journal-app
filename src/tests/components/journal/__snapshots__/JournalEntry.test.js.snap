// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <JournalEntry /> debe mostrar el snapshot correctamente  1`] = `
<div
  className="journal__entry pointer animate__animated animate__backInLeft animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://res.cloudinary.com/dhc6rsfox/image/upload/v1596500035/axoj3ozqkbfxxdch6tn1.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="journal__entry-title"
    >
      Solfeo
    </p>
    <p
      className="journal__entry-content"
    >
      30 minutos.
    </p>
  </div>
  <div
    className="journal__entry-date-box"
  >
    <span>
      Monday
    </span>
    <h4>
      27th
    </h4>
  </div>
</div>
`;
